### ISSUE TITLE
I seem to have missed a step somewhere.

### ISSUE THREAD

**Issue by niarbeht** [Rfvgyhn/min-ed-launcher#3]
```
$ env STEAM_COMPAT_DATA_PATH=/rpool/games/steamapps/compatdata/359320/ ./MinEdLauncher /rpool/games/steamapps/common/Elite\ Dangerous/EDLaunch.exe /autoquit /EDH
[16:22:08 INF] Elite Dangerous - Minimal Launcher
Platform: Dev
OS: Linux64
CobraBay Version: 0.4.6600.0
Launcher Version: 0.1.1
Products Dir: /rpool/games/steamapps/common/Elite Dangerous/Products
[16:22:09 ERR] Unhandled exception: System.AggregateException: AggregateException_ctor_DefaultMessage (Arg_NotImplementedException)
 ---> System.NotImplementedException: Arg_NotImplementedException
   at MinEdLauncher.Api.authenticate@73.Invoke(Unit unitVar)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   at MinEdLauncher.App.authenticate@19-17.Invoke(Unit unitVar)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   at MinEdLauncher.App.login@36.Invoke(Unit unitVar)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   at MinEdLauncher.App.continuation@1-23(LauncherSettings settings, String productsDir, HttpClient _arg1, DateTime localTime, Int64 _arg3, FSharpFunc`2 runningTime, FSharpResult`2 _arg4)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr callersRetAddrSlot, IntPtr callTarget, IntPtr retVal)
   at MinEdLauncher.App.continuation@1-22(LauncherSettings settings, String productsDir, HttpClient _arg1, DateTime localTime, Int64 _arg3)
   at Ply.TplPrimitives.AwaitableContinuation`3.Invoke(Unit r)
   at Ply.TplPrimitives.tryFinally@257.Invoke(AwaitableContinuation`3 this)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   at MinEdLauncher.App.run@99-6.Invoke(Unit unitVar0)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   Exception_EndOfInnerExceptionStack
```
No joy.  What other information might be useful?

**Comment by Rfvgyhn** [comment]
So it looks like it didn't recognize it was run from steam (`Platform: Dev`). Can you provide the logs from `/rpool/games/steamapps/common/Elite\ Dangerous/logs/min-ed-launcher.log`?

As a temporary workaround, try adding the `/steam` arg to your launch options.

**Comment by niarbeht** [comment]
Okay, sorry, I left some stuff out because I was aggravated by E:D's mats grind yesterday.

To start, if I place `gnome-terminal -- MinEdLauncher %command% /autoquit /EDH` as my launch options in Steam, I get

```
There was an error creating the child process for this terminal

Failed to execute child process “MinEdLauncher”: Failed to execve: No such file or directory
```

as error output from gnome-terminal.  I can get a step farther by putting `gnome-terminal -- "/rpool/games/steamapps/common/Elite Dangerous/MinEdLauncher" %command% /autoquit /EDH` as my launch options, but then I just get the good ol' "ELFCLASS32" message in gnome-terminal very briefly before it closes.  The ELFCLASS32 message can be safely ignored.  The log now shows:

```
2020-12-12 07:07:37.565 -06:00 [DBG] Args: [|"/rpool/games/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point";
  "--deploy=soldier"; "--suite=soldier"; "--verb=waitforexitandrun"; "--";
  "/rpool/games/steamapps/common/Proton 5.13/proton"; "waitforexitandrun";
  "/rpool/games/steamapps/common/Elite Dangerous/EDLaunch.exe"; "/EDH";
  "/steamid"; "441340"; "/novr"; "/autoquit"; "/EDH"|]
2020-12-12 07:07:37.673 -06:00 [DBG] Reading settings from '%XDG_CONFIG_HOME%/min-ed-launcher/settings.json'
2020-12-12 07:07:37.957 -06:00 [DBG] Settings: Error "Failed to find Elite Dangerous install directory"
2020-12-12 07:07:37.958 -06:00 [ERR] Failed to find Elite Dangerous install directory
```

I'm not sure what to make of things at this point.

If I then go and use the args from the logs at a terminal, we get:

`$ env STEAM_COMPAT_DATA_PATH=/rpool/games/steamapps/compatdata/359320/ ./MinEdLauncher "/rpool/games/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point" "--deploy=soldier" "--suite=soldier" "--verb=waitforexitandrun" "--" "/rpool/games/steamapps/common/Proton 5.13/proton" "waitforexitandrun" "/rpool/games/steamapps/common/Elite Dangerous/EDLaunch.exe" "/autoquit" "/EDH" "/steam"`

which might be some of the weirdest soup I've ever seen.  If we get rid of everything before the `"--"`, we get:

`$ env STEAM_COMPAT_DATA_PATH=/rpool/games/steamapps/compatdata/359320/ ./MinEdLauncher "/rpool/games/steamapps/common/Proton 5.13/proton" "waitforexitandrun" "/rpool/games/steamapps/common/Elite Dangerous/EDLaunch.exe" "/autoquit" "/EDH" "/steam"`

with output:

```
[07:13:03 INF] Elite Dangerous - Minimal Launcher
Platform: Steam
OS: Linux64
CobraBay Version: 0.4.6600.0
Launcher Version: 0.1.1
Products Dir: /rpool/games/steamapps/common/Elite Dangerous/Products
[07:13:04 ERR] Couldn't login: Unable to find steam library 'libsteam_api.so'
```

I really have no idea.

One thing to note is that I am using the ED Mono ( https://github.com/redmcg/wine-mono/wiki ), so maybe that's interfering with something?  I tried uninstalling it, the behavior remained the same.  I suspect that the uninstaller might just be a shim, so I might have to nuke the wine prefix if it's ED Mono doing it.

**Comment by niarbeht** [comment]
For the sake of hilarity I tried `env STEAM_COMPAT_DATA_PATH=/rpool/games/steamapps/compatdata/359320/ "/rpool/games/steamapps/common/Proton 5.13/proton" "waitforexitandrun" "/rpool/games/steamapps/common/Elite Dangerous/MinEdLauncher.exe" "Z:/rpool/games/steamapps/common/Elite Dangerous/EDLaunch.exe" "/autoquit" "/EDH" "/steam"` just to see what would happen.

```
2020-12-12 07:25:48.206 -06:00 [DBG] Args: [|"Z:/rpool/games/steamapps/common/Elite Dangerous/EDLaunch.exe"; "/autoquit";
  "/EDH"; "/steam"|]
2020-12-12 07:25:48.278 -06:00 [DBG] Reading settings from 'C:\users\steamuser\Local Settings\Application Data\min-ed-launcher\settings.json'
2020-12-12 07:25:48.505 -06:00 [DBG] Settings: Ok { Platform = Steam
     DisplayMode = Pancake
     AutoRun = false
     AutoQuit = true
     WatchForCrashes = false
     ProductWhitelist = set ["edh"]
     ForceLocal = true
     Proton = None
     CbLauncherDir = "Z:\rpool\games\steamapps\common\Elite Dangerous"
     ApiUri = https://api.zaonce.net/
     Restart = (false, 3000L)
     Processes = [] }
2020-12-12 07:25:48.515 -06:00 [INF] Elite Dangerous - Minimal Launcher
Platform: Steam
OS: Win64
CobraBay Version: 0.4.6600.0
Launcher Version: 0.1.1
Products Dir: Z:\rpool\games\steamapps\common\Elite Dangerous\Products
2020-12-12 07:25:49.449 -06:00 [DBG] Disposing Steam resources
2020-12-12 07:25:49.451 -06:00 [ERR] Couldn't login: Unable to initialize Steam
```

So, that was the craziest thing I could think to try.  Sometimes going full-on crazy person works.  This time, no dice.

**Comment by Rfvgyhn** [comment]
> I was aggravated by E:D's mats grind yesterday
I certainly know that feeling.

tldr; Is MinEdLauncher in the same dir as `EDLaunch.exe`. If not, place it there. If so, does forcing steam to use proton 5.0-10 for elite change anything?

`Failed to execute child process “MinEdLauncher”: Failed to execve: No such file or directory` and `Couldn't login: Unable to find steam library 'libsteam_api.so` sounds like `MinEdLauncher` isn't in the same directory as `EDLaunch.exe`. 

`Reading settings from '%XDG_CONFIG_HOME%/min-ed-launcher/settings.json'` Environment variable didn't get expanded here.

It definitely seems like something weird is going on with the environment. I haven't seen the `SteamLinuxRuntime_soldier` + other soldier references before. That looks like proton is doing something new. Does anything change for you if tell steam to use proton 5.0-10 for elite?

Looking at https://github.com/ValveSoftware/Proton/issues/4269, it seems like people are having issues with proton 5.13. When I get to a machine on which I can test, I'll try forcing that version and see if I get any of the issues you're having.

Regarding ED Mono, I haven't used it before, but I _think_ it shouldn't be needed. As far as I know, only the default launcher requires .net. The game itself doesn't and the min launcher has a stripped down version embedded in it. Even if it were installed though, I wouldn't think it would interfere with things since wine isn't needed/used to launch the min launcher.

**Comment by Rfvgyhn** [comment]
Ok, I had a chance to run with proton 5.13 (not sure why steam didn't automatically update me to 5.13) and I'm seeing similar problems (Failed to find Elite Dangerous install directory). Proton 5.13 runs via pressure-vessel and I'll need to add support for that.

**Comment by niarbeht** [comment]
> Proton 5.13 runs via pressure-vessel and I'll need to add support for that.

I'll just wait, then.  Thanks for helping with this!

**Comment by Rfvgyhn** [comment]
Not sure how long it will take to fix this. It's simple enough to update the launcher to recognize the new arguments but I'm unable to get any game, let alone Elite, to run with proton 5.13 so I can't test anything.

Temporary workaround is to run Elite with proton 5.0.

**Comment by Rfvgyhn** [comment]
After some combo of reinstalling and verifying proton 5.13 and steam runtime linux, I got proton 5.13 working. Can you test the [latest CI build](https://github.com/Rfvgyhn/min-ed-launcher/actions/runs/418121400) and see if things work for you? Elite crashes with proton 5.13 for me but it does launch.

**Comment by niarbeht** [comment]
Same behavior with the CI build.

I'm gonna see if changing my default shell to bash and logging back in fixes anything.

**Comment by Rfvgyhn** [comment]
Same behavior as which one of your attempts? Just the regular launch via steam with options `gnome-terminal -- MinEdLauncher %command% /autoquit /EDH`? If so, I do have a typo in the readme. The steam launch options should be `gnome-terminal -e ./MinEdLauncher %command% /autoquit /EDH`. I forgot the current directory `./` prefix on the exe.

**Comment by niarbeht** [comment]
With `gnome-terminal -- ./MinEdLauncher %command% /autoquit /EDH`, I get the following:

``` 
2020-12-13 11:28:44.662 -06:00 [DBG] Args: [|"/rpool/games/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point";
  "--deploy=soldier"; "--suite=soldier"; "--verb=waitforexitandrun"; "--";
  "/rpool/games/steamapps/common/Proton 5.13/proton"; "waitforexitandrun";
  "/rpool/games/steamapps/common/Elite Dangerous/EDLaunch.exe"; "/EDH";
  "/steamid"; "441340"; "/novr"; "/autoquit"; "/EDH"|]
2020-12-13 11:28:44.779 -06:00 [DBG] Reading settings from '%XDG_CONFIG_HOME%/min-ed-launcher/settings.json'
2020-12-13 11:28:45.112 -06:00 [DBG] Settings: Ok
  { Platform = Steam
    DisplayMode = Pancake
    AutoRun = false
    AutoQuit = true
    WatchForCrashes = false
    ProductWhitelist = set ["edh"]
    ForceLocal = true
    Proton =
            Some
              { EntryPoint =
                            "/rpool/games/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point"
                Args =
                      [|"--deploy=soldier"; "--suite=soldier";
                        "--verb=waitforexitandrun"; "--"; "python3";
                        "/rpool/games/steamapps/common/Proton 5.13/proton";
                        "waitforexitandrun"|] }
    CbLauncherDir = "/rpool/games/steamapps/common/Elite Dangerous"
    ApiUri = https://api.zaonce.net/
    Restart = (false, 3000L)
    Processes = [] }
2020-12-13 11:28:45.163 -06:00 [INF] Elite Dangerous - Minimal Launcher
Platform: Steam
OS: Linux64
CobraBay Version: 0.4.6600.0
Launcher Version: 0.1.1-ci418121400
Products Dir: /rpool/games/steamapps/common/Elite Dangerous/Products
2020-12-13 11:28:46.193 -06:00 [DBG] Disposing Steam resources
2020-12-13 11:28:46.196 -06:00 [ERR] Couldn't login: Unable to find steam library 'libsteam_api.so'
```

So, progress.  Big progress.

I tried linking the `libsteam_api.so` file used by proton 5.13 into the Elite Dangerous directory, and the output changed to this on run:

```
2020-12-13 11:32:21.190 -06:00 [DBG] Args: [|"/rpool/games/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point";
  "--deploy=soldier"; "--suite=soldier"; "--verb=waitforexitandrun"; "--";
  "/rpool/games/steamapps/common/Proton 5.13/proton"; "waitforexitandrun";
  "/rpool/games/steamapps/common/Elite Dangerous/EDLaunch.exe"; "/EDH";
  "/steamid"; "441340"; "/novr"; "/autoquit"; "/EDH"|]
2020-12-13 11:32:21.309 -06:00 [DBG] Reading settings from '%XDG_CONFIG_HOME%/min-ed-launcher/settings.json'
2020-12-13 11:32:21.641 -06:00 [DBG] Settings: Ok
  { Platform = Steam
    DisplayMode = Pancake
    AutoRun = false
    AutoQuit = true
    WatchForCrashes = false
    ProductWhitelist = set ["edh"]
    ForceLocal = true
    Proton =
            Some
              { EntryPoint =
                            "/rpool/games/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point"
                Args =
                      [|"--deploy=soldier"; "--suite=soldier";
                        "--verb=waitforexitandrun"; "--"; "python3";
                        "/rpool/games/steamapps/common/Proton 5.13/proton";
                        "waitforexitandrun"|] }
    CbLauncherDir = "/rpool/games/steamapps/common/Elite Dangerous"
    ApiUri = https://api.zaonce.net/
    Restart = (false, 3000L)
    Processes = [] }
2020-12-13 11:32:21.693 -06:00 [INF] Elite Dangerous - Minimal Launcher
Platform: Steam
OS: Linux64
CobraBay Version: 0.4.6600.0
Launcher Version: 0.1.1-ci418121400
Products Dir: /rpool/games/steamapps/common/Elite Dangerous/Products
2020-12-13 11:32:22.783 -06:00 [DBG] Disposing Steam resources
2020-12-13 11:32:22.818 -06:00 [ERR] Unhandled exception: System.AggregateException: AggregateException_ctor_DefaultMessage (Unable to load shared library 'libsteam_api.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibsteam_api.so: cannot open shared object file: No such file or directory)
 ---> System.DllNotFoundException: Unable to load shared library 'libsteam_api.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibsteam_api.so: cannot open shared object file: No such file or directory
   at MinEdLauncher.Steam.SteamAPI_Init()
   at MinEdLauncher.Steam.Steam.init()
   at MinEdLauncher.Steam.Steam.Login()
   at MinEdLauncher.App.body@1-16(FSharpFunc`2 runningTime, HttpClient httpClient, String machineId, Platform platform, IDisposable noopDisposable, Steam _arg4)
   at MinEdLauncher.App.login@39-3.Invoke(Unit unitVar0)
   at Ply.TplPrimitives.tryFinally[u](FSharpFunc`2 continuation, FSharpFunc`2 finallyBody)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   at MinEdLauncher.App.continuation@1-23(LauncherSettings settings, String productsDir, HttpClient _arg1, DateTime localTime, Int64 _arg3, FSharpFunc`2 runningTime, FSharpResult`2 _arg4)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr callersRetAddrSlot, IntPtr callTarget, IntPtr retVal)
   at MinEdLauncher.App.continuation@1-22(LauncherSettings settings, String productsDir, HttpClient _arg1, DateTime localTime, Int64 _arg3)
   at Ply.TplPrimitives.AwaitableContinuation`3.Invoke(Unit r)
   at Ply.TplPrimitives.tryFinally@257.Invoke(AwaitableContinuation`3 this)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   at MinEdLauncher.App.run@99-6.Invoke(Unit unitVar0)
   at Ply.TplPrimitives.ContinuationStateMachine`1.System-Runtime-CompilerServices-IAsyncStateMachine-MoveNext()
   Exception_EndOfInnerExceptionStack
```

I'm running Steam in native mode instead of runtime mode, so maybe switching to runtime mode will help.  I'll go test that next.

**Comment by niarbeht** [comment]
Starting Steam in runtime mode means it can't find `/usr/bin/gnome-terminal` for some reason, so I'm just having it run MinEdLauncher directly and seeing how that does.

...

Aaaaaaaaaaand making a symlink to the `libsteam_api.so` used by Factorio makes it work.  MYSTERIOUS.

**Comment by Rfvgyhn** [comment]
Looks like another oversight on my part. I forgot to bundle the steam api lib. Unfortunately, the proton `libsteam_api.so` is 32bit and the launcher is 64bit so it can't load it properly. [Latest CI build](https://github.com/Rfvgyhn/min-ed-launcher/actions/runs/419580210) should have fixed that.

No idea why it wouldn't be able to find gnome-terminal. I'm using runtime mode and I don't have any issues with alacritty. Just note that if you don't run the launcher in its own terminal, it will end up crashing if you enable the auto restart mode since there's no input that can be read.

If the latest build works without you having to symlink/copy `libsteam_api.so` from somewhere else, I'll release a new version with the fixes.

Thanks for taking the time to debug all of this.

**Comment by niarbeht** [comment]
I'll look later, I'm hauling atm :P

**Comment by niarbeht** [comment]
Okay, tried the new CI build with gnome-terminal, I got `failed to execve blah blah something about permissions`, turns out the execute flag wasn't set on the binary.

I'm a glass of wine deep and I forgot to get the exact error message, but that's the cause.
